cmd/util/ledger/migrations/cadence_data_validation.go

```diff
@@ -0,0 +1,329 @@
+
+func (m *preMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	m.log = log.With().Str("component", "CadenceDataValidationPreMigration").Logger()
+
+	return nil
+}
+
+func (m *preMigration) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+
+	accountHash, err := m.v.hashAccountCadenceValues(address, payloads)
+	if err != nil {
+		m.log.Error().
+			Err(err).
+			Hex("address", address[:]).
+			Msg("failed to hash cadence values")
+
+		// on error still continue with the migration
+		return payloads, nil
+	}
+
+	m.v.setAccountHash(address, accountHash)
+
+	return payloads, nil
+}
+
+type postMigration struct {
+	log zerolog.Logger
+
+	rwf reporters.ReportWriterFactory
+	rw  reporters.ReportWriter
+
+	v *CadenceDataValidationMigrations
+}
+
+var _ AccountBasedMigration = &postMigration{}
+
+func (m *postMigration) Close() error {
+	for address := range m.v.accountHashes {
+		m.log.Error().
+			Hex("address", address[:]).
+			Msg("cadence values missing after migration")
+
+		m.rw.Write(
+			cadenceDataValidationReportEntry{
+
+				Address: address.Hex(),
+				Problem: "cadence values missing after migration",
+			},
+		)
+	}
+	return nil
+}
+
+func (m *postMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	m.log = log.With().Str("component", "CadenceDataValidationPostMigration").Logger()
+	m.rw = m.rwf.ReportWriter("cadence_data_validation")
+
+	return nil
+}
+
+func (m *postMigration) MigrateAccount(
+	ctx context.Context,
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+	newHash, err := m.v.hashAccountCadenceValues(address, payloads)
+	if err != nil {
+		m.log.Info().
+			Err(err).
+			Hex("address", address[:]).
+			Msg("failed to hash cadence values")
+		return payloads, nil
+	}
+
+	accountHash, ok := m.v.getAccountHash(address)
+
+	if !ok {
+		m.log.Error().
+			Hex("address", address[:]).
+			Msg("cadence values missing before migration")
+
+		m.rw.Write(
+			cadenceDataValidationReportEntry{
+
+				Address: address.Hex(),
+				Problem: "cadence values missing before migration",
+			},
+		)
+	}
+	if !bytes.Equal(accountHash, newHash) {
+		m.log.Error().
+			Hex("address", address[:]).
+			Msg("cadence values mismatch")
+
+		m.rw.Write(
+			cadenceDataValidationReportEntry{
+
+				Address: address.Hex(),
+				Problem: "cadence values mismatch",
+			},
+		)
+	}
+
+	// remove the address from the map so we can check if there are any
+	// missing addresses
+	m.v.deleteAccountHash(address)
+
+	return payloads, nil
+}
+
+func (m *CadenceDataValidationMigrations) hashAccountCadenceValues(
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]byte, error) {
+	hasher := newHasher()
+	mr, err := newMigratorRuntime(address, payloads)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, domain := range domains {
```

Comment:
Can we maybe assert that the account does not have any data stored outside of those domains?

Reply:
Would it be sufficient to log payloads that are not migrated (rather than assert)?

We already log payloads that are not migrated in `AtreeRegisterMigrator.validateChangesAndCreateNewRegisters()`.

Also, `validateChangesAndCreateNewRegisters()` is before validation.  Validation only compares Cadences values in migrated domains.

https://github.com/onflow/flow-go/blob/759ba39ad83b50ab144c794381ca89e6abf54c94/cmd/util/ledger/migrations/atree_register_migration.go#L331-L337

Thoughts?

Reply:
Sounds good!